service/history/outbound_queue_standby_task_executor.go

```diff
@@ -75,14 +97,188 @@ func (e *outboundQueueStandbyTaskExecutor) Execute(
 		metrics.OperationTag(taskType),
 	}
 
-	if err == nil {
-		smRegistry := e.shardContext.StateMachineRegistry()
-		err = smRegistry.ExecuteStandbyImmediateTask(ctx, e, ref, smt)
+	if err != nil {
+		return queues.ExecuteResponse{
+			ExecutionMetricTags: metricsTags,
+			ExecutedAsActive:    false,
+			ExecutionErr:        err,
+		}
+	}
+
+	actionFn := func(
+		_ context.Context,
+		_ workflow.Context,
+		mutableState workflow.MutableState,
+	) (any, error) {
+		node, err := mutableState.HSM().Child(ref.StateMachinePath())
+		if err != nil {
+			return nil, err
+		}
+		if err := e.validateStateMachineRef(mutableState, ref); err != nil {
+			if errors.Is(err, consts.ErrStaleReference) {
+				return nil, nil
+			}
+			return nil, err
+		}
+		if smt.Concurrent() {
+			//nolint:revive // concurrent tasks implements hsm.ConcurrentTask interface
+			concurrentSmt := smt.(hsm.ConcurrentTask)
+			if err := concurrentSmt.Validate(node); err != nil {
+				if errors.Is(err, consts.ErrStaleReference) {
+					return nil, nil
+				}
+				return nil, err
+			}
+		}
+		return getHistoryResendInfo(mutableState)
 	}
 
+	err = e.processTask(
+		ctx,
+		task,
+		ref,
+		actionFn,
+		getStandbyPostActionFn(
+			task,
+			e.getCurrentTime,
+			e.config.StandbyTaskMissingEventsResendDelay(task.GetType()),
+			e.config.StandbyTaskMissingEventsDiscardDelay(task.GetType()),
+			e.fetchHistoryFromRemote,
+			standbyTimerTaskPostActionTaskDiscarded,
+		),
+	)
+
 	return queues.ExecuteResponse{
 		ExecutionMetricTags: metricsTags,
 		ExecutedAsActive:    false,
 		ExecutionErr:        err,
 	}
 }
+
+func (e *outboundQueueStandbyTaskExecutor) processTask(
+	ctx context.Context,
+	task tasks.Task,
+	ref hsm.Ref,
+	actionFn standbyActionFn,
+	postActionFn standbyPostActionFn,
+) (retError error) {
+	ctx, cancel := context.WithTimeout(ctx, taskTimeout)
+	defer cancel()
+
+	nsRecord, err := e.shardContext.GetNamespaceRegistry().GetNamespaceByID(
+		namespace.ID(task.GetNamespaceID()),
+	)
+	if err != nil {
+		return err
+	}
+	if !nsRecord.IsOnCluster(e.clusterName) {
+		// namespace is not replicated to local cluster, ignore corresponding tasks
+		return nil
+	}
+
+	executionContext, release, mutableState, err := e.getMutableState(ctx, ref.WorkflowKey)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if retError == consts.ErrTaskRetry {
+			release(nil)
+		} else {
+			release(retError)
+		}
+	}()
+
+	if mutableState == nil {
+		return nil
+	}
+
+	historyResendInfo, err := actionFn(ctx, executionContext, mutableState)
```

Comment:
Before this, please verify the task against shard clock to ensure that we don't have a stale mutable state.
@yycptt will move this validation to a higher level.